src/Orleans/Serialization/TypeUtilities.cs

```diff
@@ -20,54 +21,63 @@ internal static bool IsOrleansPrimitive(this Type t)
             return typeInfo.IsPrimitive || typeInfo.IsEnum || t == typeof(string) || t == typeof(DateTime) || t == typeof(Decimal) || (typeInfo.IsArray && typeInfo.GetElementType().IsOrleansPrimitive());
         }
 
-        static readonly Dictionary<RuntimeTypeHandle, bool> shallowCopyableValueTypes = new Dictionary<RuntimeTypeHandle, bool>();
+        static readonly ConcurrentDictionary<Type, bool> shallowCopyableTypes = new ConcurrentDictionary<Type, bool>();
         static readonly Dictionary<RuntimeTypeHandle, string> typeNameCache = new Dictionary<RuntimeTypeHandle, string>();
         static readonly Dictionary<RuntimeTypeHandle, string> typeKeyStringCache = new Dictionary<RuntimeTypeHandle, string>();
         static readonly Dictionary<RuntimeTypeHandle, byte[]> typeKeyCache = new Dictionary<RuntimeTypeHandle, byte[]>();
 
         static TypeUtilities()
         {
-            shallowCopyableValueTypes[typeof(Decimal).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(DateTime).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(TimeSpan).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(IPAddress).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(IPEndPoint).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(SiloAddress).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(GrainId).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(ActivationId).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(ActivationAddress).TypeHandle] = true;
-            shallowCopyableValueTypes[typeof(CorrelationId).TypeHandle] = true;
+            shallowCopyableTypes[typeof(Decimal)] = true;
+            shallowCopyableTypes[typeof(DateTime)] = true;
+            shallowCopyableTypes[typeof(TimeSpan)] = true;
+            shallowCopyableTypes[typeof(IPAddress)] = true;
+            shallowCopyableTypes[typeof(IPEndPoint)] = true;
+            shallowCopyableTypes[typeof(SiloAddress)] = true;
+            shallowCopyableTypes[typeof(GrainId)] = true;
+            shallowCopyableTypes[typeof(ActivationId)] = true;
+            shallowCopyableTypes[typeof(ActivationAddress)] = true;
+            shallowCopyableTypes[typeof(CorrelationId)] = true;
+            shallowCopyableTypes[typeof(string)] = true;
+            shallowCopyableTypes[typeof(Immutable<>)] = true;
+            shallowCopyableTypes[typeof(CancellationToken)] = true;
         }
 
         internal static bool IsOrleansShallowCopyable(this Type t)
         {
+            bool result;
+            if (shallowCopyableTypes.TryGetValue(t, out result))
```

Comment:
Ohhh, I see now. It will do it automatically, after the first check.
So this is probably the root of the majority of the optimization:

for shallow copiabales:
After a couple of iterations we stop mutating the `ConcurrentDictionary`, and then most of the times we hit the Dictionary and exit quickly. While in the old code we did `GetTypeInfo()` and `IsPrimitive` and all the other check.

For non shallow ones: we basically do the same set of checks as before, so its not worth.